core/controllers/diagnostic_test_player.py

```diff
@@ -103,17 +97,38 @@ def get(self, topic_id: str) -> None:
         for question in questions:
             linked_skill_ids = question.linked_skill_ids
             diagnostic_test_linked_skill_ids = list(
-                set(linked_skill_ids).intersection(
-                    set(diagnostic_test_skill_ids))
-            )
+                set(linked_skill_ids) & set(diagnostic_test_skill_ids))
 
             for skill_id in diagnostic_test_linked_skill_ids:
-                if 'main_question' not in skill_id_to_questions_dict[skill_id]:
-                    skill_id_to_questions_dict[skill_id]['main_question'] = (
-                        question.to_dict())
-                else:
-                    skill_id_to_questions_dict[skill_id]['backup_question'] = (
-                        question.to_dict())
+                skill_id_to_questions_map[skill_id].append(question)
```

Comment:
Logically, does it make sense for the same question to be linked to multiple skills in the same diagnostic test? Wouldn't that cause the same question to show up twice in the test?

Reply:
Yes, I understand this point. I have made some changes and need some suggestions from your side. Please take a look at the following points:

- The method [get_questions_by_skill_ids()](https://github.com/oppia/oppia/pull/16408/files#diff-6c73058908aa0dc0c49902d0897c81a6b2911a6c659f927dbea641124ba56fa9R91) distributes the # questions equally to all the skills (diagnostic test skills in this case).
- If any question is linked to two diagnostic test skills, then the total fetched question should contain repetition of questions.
- [In line 99](https://github.com/oppia/oppia/pull/16408/files#diff-6c73058908aa0dc0c49902d0897c81a6b2911a6c659f927dbea641124ba56fa9R99), I have checked the repeated question should not be considered.
- Now if the repeated question is ignored then there must be 1 or more skill which has less than two questions (not satisfying the condition for the main question and the backup question)
- In this case, if we raise an exception (which I was doing earlier), this does not help the learner in any way, thus I have decided to skip that skill for testing ([line 126](https://github.com/oppia/oppia/pull/16408/files#diff-6c73058908aa0dc0c49902d0897c81a6b2911a6c659f927dbea641124ba56fa9R126))

Please provide your feedback to this.
Thank you.

Reply:
Hm, I think that the check in line 99 is problematic. Normally when checking equality, try to use something primitive like a string, not a dict. Object equality checking has a bunch of pitfalls. Instead, use something like the question_id (you can create a skill_id to question_id map from the get-go).

For the more general issue of duplication, that is indeed a bit tricky. Maybe what you can do is, if you find that a question tests multiple skills, then you could either (a) ask it twice in the frontend, or (b) ask it once, but remember the result and apply it to the other skill when the time comes. I feel like (a) is actually better in this case because the code will be a lot simpler and this case is unlikely to happen often.

Whatever you decide, please document the rationale here. Thanks!

Reply:
Oh, one other idea that might be more robust. You could initially look at the questions and see if any cover more than one of the given skills. If so, ignore that question.

Then, you'll get a skill ID to questions mapping that will have some "holes". Find the skills corresponding these holes and do another independent fetch of questions to fill these holes (possibly one question at a time). I think that that would be foolproof and reasonably efficient (since in the general case the additional fetches won't be needed). What do you think?

Reply:
Thank you for suggesting different approaches: 
Re the last approach: 
Here, we are thinking in the context of a single topic, and in a topic, we have different diagnostic test skills, which may have the same linked questions.

[Assumptions]: In a classroom, we have different topics, and 2 or more topics contain the same diagnostic test skill(s). So in that case again there is a chance of repeating the questions because we are encountering one topic at a time.

So I am thinking of it in such a way, that what if we present the question even if they are repeated (as per suggestion 1) and in the later stage of the project update the things as per suggestions from different stakeholders?

Please provide your view regarding this.

_[The comment is for clarification, no changes were made after the last review.]_

Reply:
Ah sorry, I think I get it, you are fetching questions from the backend for one topic at a time, right? Yes I think your suggestion is fine then, thanks.

Reply:
In fact, that simplifies things, right? You could just pass the IDs of the questions you've already fetched (in e.g. an "exclude_question_ids" parameter) to the backend, and have the backend logic omit those question IDs when computing what to return. E.g. do the fetch as normal and if it hits a duplicate then do it again for those skills (but fetch a bunch of questions and pick the first non-duplicate one).

Reply:
Thank you @seanlip for suggesting this approach. 
I have updated the implementation such that the handler accepts the exclude_question_ids list. I think the test coverage will miss but after accepting the changes (if any) in the logic I will add some tests.
Please take a look once you are free.

Thank you.